packages/lesswrong/lib/vulcan-lib/fragments.ts

```diff
@@ -151,6 +157,169 @@ const getFragmentText = (fragmentName: FragmentName): string => {
+ * a good way of checking the "intended" type of any given field inside one of those fragments.
+ */
+const validateFragmentSelections = (fragmentDef: FragmentDefinitionNode | FieldNode, typeName: string) => {
+  const selectionSet = fragmentDef.selectionSet;
+  
+  // Get the collection for this type if it exists
+  let collection: CollectionBase<CollectionNameString>|null = null;
+  try {
+    collection = getCollectionByTypeName(typeName.replace(/[![\]]+/g, ""));
+  } catch (e) {
+    // Not a collection type, skip validation
+    return;
+  }
+
+  if (!selectionSet) return;
+
+  const schema = collection._schemaFields;
+
+  // For fragments on collections, we only need to check fields that have a resolver
+  // Other fields won't have _id fields nested in them (with the exception of normalized editable fields, which get treated as resolver fields anyways)
+  const resolverFields = Object.keys(schema).filter(fieldName => schema[fieldName].resolveAs);
+
+  // However, since we'll be mapping over the selection set, which will contain field names that don't actually exist in the schema,
+  // we need to map the "resolver field" names referenced in the fragment to the actual schema field names that generated them
+  const resolverFieldToOriginalFieldMap = Object.fromEntries(
+    resolverFields.map(fieldName => [schema[fieldName].resolveAs?.fieldName ?? fieldName, fieldName])
+  );
+
+  const selectionResolverFieldMetadata = selectionSet.selections
+    .filter((selection): selection is FieldNode => selection.kind === 'Field')
+    .map(selection => {
+      const { field, fieldName } = resolveFragmentFieldNode(selection, resolverFieldToOriginalFieldMap, schema);
+      return { field, fieldName, selection };
+    })
+    // Filter out fields that don't have a resolver
+    .filter(({ field }) => !!field?.resolveAs)
+    .map(({ fieldName, field, selection }) => {
+      const fieldType = field.resolveAs?.type ?? getGraphQLType(schema, fieldName);
+      return { fieldName, fieldType, selection };
+    })
+    // Filter out fields that aren't even possibly collection types
+    .filter(isMaybeCollectionFieldInfo);
+
+  for (const { selection, fieldName, fieldType } of selectionResolverFieldMetadata) {
+    const validCollectionName = isValidCollectionName(fieldType);
+    
+    if (validCollectionName && selection.selectionSet) {
+      const { selections } = selection.selectionSet;
+
+      // Check if there's a fragment spread - if so, we don't need to check for _id
+      // since we'll check the nested fragment at some point, and if that doesn't have _id,
+      // we'll fail validation anyway.
+      const hasSpread = selections.some(s => s.kind === 'FragmentSpread');
+      const hasId = selections.some(s => s.kind === 'Field' && s.name.value === '_id');
+
+      // If there's no fragment spread and no _id field on a field that's resolved to a collection type, throw an error
+      if (!hasSpread && !hasId) {
```

Comment:
We're checking this for subfields of fragments, but not for top-level fields in fragments, which probably we should do? For instance, I can remove the `_id` field from `PostsMinimumInfo` and this check passes successfully.